File: frontend/src/types/game.ts
export type Suit = 'spades' | 'hearts' | 'diamonds' | 'clubs';
export type Rank = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13;

export interface Card {
  suit: Suit;
  rank: Rank;
}

export type AceDirection = 'undecided' | 'low' | 'high';
export type GameStatus = 'waiting' | 'playing' | 'finished';
export type MoveType = 'play' | 'face_down';

export interface SuitSequence {
  suit: Suit;
  low: number;
  high: number;
  closed: boolean;
}

export interface PlayerResult {
  seat: number;
  penalty_points: number;
  rank: number;
  face_down_cards: Card[];
}

export interface Move {
  seat: number;
  move_num: number;
  type: MoveType;
  card: Card;
}

export interface PlayerView {
  seat: number;
  user_id: string;
  display_name: string;
  card_count: number;
  face_down_count: number;
  is_bot: boolean;
  is_current_turn: boolean;
}

export interface GameState {
  hand: Card[];
  sequences: SuitSequence[];
  players: PlayerView[];
  current_turn: number;
  ace_direction: AceDirection;
  status: GameStatus;
  valid_moves: Card[];
  must_face_down: boolean;
  move_history: Move[];
  results?: PlayerResult[];
  my_seat: number;
}

export interface User {
  id: string;
  display_name: string;
  email: string;
  avatar_url: string;
  elo_rating: number;
  stats?: PlayerStats;
}

export interface PlayerStats {
  elo_rating: number;
  total_games: number;
  wins: number;
  losses: number;
  total_penalty_points: number;
}

export interface Room {
  id: string;
  code: string;
  host_user_id: string;
  status: string;
  bot_enabled: boolean;
  bot_count: number;
  turn_timer: number;
  players: RoomPlayer[];
}

export interface RoomPlayer {
  user_id: string;
  seat: number;
  is_bot: boolean;
  bot_difficulty: string;
  user?: User;
}

export interface LeaderboardEntry {
  user_id: string;
  display_name: string;
  elo_rating: number;
  total_games: number;
  wins: number;
}

File: backend/domain/game.go
package domain

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// --- Game ---

type GameDBStatus string

const (
	GameDBStatusInProgress GameDBStatus = "in_progress"
	GameDBStatusFinished   GameDBStatus = "finished"
)

type Game struct {
	ID           uuid.UUID      `json:"id"            gorm:"type:uuid;primarykey"`
	RoomID       uuid.UUID      `json:"room_id"       gorm:"type:uuid;index"`
	AceDirection string         `json:"ace_direction"` // undecided, low, high
	Status       GameDBStatus   `json:"status"        gorm:"default:in_progress"`
	StartedAt    time.Time      `json:"started_at"`
	EndedAt      *time.Time     `json:"ended_at"`
	CreatedAt    time.Time      `json:"created_at"`
	UpdatedAt    time.Time      `json:"updated_at"`
	DeletedAt    gorm.DeletedAt `json:"-"             gorm:"index"`

	// Relations
	Players []GamePlayer `json:"players,omitempty" gorm:"foreignKey:GameID"`
	Moves   []GameMove   `json:"moves,omitempty"   gorm:"foreignKey:GameID"`
}

type GamePlayer struct {
	ID            uuid.UUID `json:"id"             gorm:"type:uuid;primarykey"`
	GameID        uuid.UUID `json:"game_id"        gorm:"type:uuid;index"`
	UserID        uuid.UUID `json:"user_id"        gorm:"type:uuid;index"`
	Seat          int       `json:"seat"`
	IsBot         bool      `json:"is_bot"         gorm:"default:false"`
	BotDifficulty string    `json:"bot_difficulty"`
	PenaltyPoints int       `json:"penalty_points" gorm:"default:0"`
	FinalRank     int       `json:"final_rank"     gorm:"default:0"`

	// Relations
	User *User `json:"user,omitempty" gorm:"foreignKey:UserID"`
}

type GameMove struct {
	ID        uuid.UUID `json:"id"         gorm:"type:uuid;primarykey"`
	GameID    uuid.UUID `json:"game_id"    gorm:"type:uuid;index"`
	Seat      int       `json:"seat"`
	MoveNum   int       `json:"move_num"`
	MoveType  string    `json:"move_type"` // play, face_down
	CardSuit  string    `json:"card_suit"`
	CardRank  int       `json:"card_rank"`
	CreatedAt time.Time `json:"created_at"`
}

// --- Player Stats ---

type PlayerStats struct {
	UserID            uuid.UUID `json:"user_id"             gorm:"type:uuid;primarykey"`
	TotalGames        int       `json:"total_games"         gorm:"default:0"`
	Wins              int       `json:"wins"                gorm:"default:0"`
	Losses            int       `json:"losses"              gorm:"default:0"`
	TotalPenaltyPoints int      `json:"total_penalty_points" gorm:"default:0"`
	EloRating         int       `json:"elo_rating"          gorm:"default:1000"`
	UpdatedAt         time.Time `json:"updated_at"`

	// Relations
	User *User `json:"user,omitempty" gorm:"foreignKey:UserID"`
}

// --- Filters ---

type GameFilter struct {
	ID     *uuid.UUID
	RoomID *uuid.UUID
	Status *GameDBStatus
}

func (f *GameFilter) Query(q *gorm.DB) {
	if f.ID != nil {
		q.Where("id = ?", *f.ID)
	}
	if f.RoomID != nil {
		q.Where("room_id = ?", *f.RoomID)
	}
	if f.Status != nil {
		q.Where("status = ?", *f.Status)
	}
}
